### PR TITLE
Cleanup `CMakeList.txt`

### PR DESCRIPTION
**Pull request by niboshi**
Organized the structure of `CMakeList.txt`

### CODE REVIEW

File: chainerx_cc/CMakeLists.txt [hunk 1]
```diff
@@ -12,49 +12,57 @@ SET(CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake ${CMAKE_MODULE_PATH})
...
-option(CHAINERX_BUILD_PYTHON "Build Python binding" OFF)
-option(CHAINERX_BUILD_TEST "Build test" OFF)
-option(CHAINERX_BUILD_EXAMPLES "Build examples" OFF)
-option(CHAINERX_WARNINGS_AS_ERRORS "Make all warnings of compilers into errors" ON)
-option(CHAINERX_ENABLE_THREAD_SANITIZER "Enable thread sanitizer." OFF)
+#---------
+# Options
+#---------
 
 # CHAINERX_CUDNN_USE_CUPY
```
Comment (take-cheeze) [inline chainerx_cc/CMakeLists.txt]: Is it possible to make this envvar checking a macro or function?

Comment (niboshi) [reply chainerx_cc/CMakeLists.txt]: Could be done, but maybe in a separate PR?

**Comment by take-cheeze** [comment]
Anyway Jenkins, test this please!

File: chainerx_cc/CMakeLists.txt [hunk 2]
```diff
@@ -65,13 +73,66 @@ if(NOT ((CMAKE_BUILD_TYPE STREQUAL "Release") OR (CMAKE_BUILD_TYPE STREQUAL "Deb
     message(FATAL_ERROR "Invalid CMAKE_BUILD_TYPE: " ${CMAKE_BUILD_TYPE})
 endif()
 
-# CUDA
-if(${CHAINERX_BUILD_CUDA})
+#-----------
+# C++ setup
+#-----------
+set(CMAKE_CXX_STANDARD 14)
```
Comment (take-cheeze) [inline chainerx_cc/CMakeLists.txt]: Seems some compilation error is caused by the move of this.

Comment (niboshi) [reply chainerx_cc/CMakeLists.txt]: Fixed (hopefully): moved `list(APPEND CUDA_NVCC_FLAGS -std=c++14)`

**Comment by niboshi** [comment]
PTAL

**Comment by take-cheeze** [comment]
Jenkins, test this please!

File: chainerx_cc/CMakeLists.txt (hunk 2 above)
Comment (take-cheeze) [reply chainerx_cc/CMakeLists.txt]: I'd like [`CUDA_STANDARD`](https://cmake.org/cmake/help/v3.8/prop_tgt/CUDA_STANDARD.html#prop_tgt:CUDA_STANDARD) more if we can use cmake with its support

File: chainerx_cc/CMakeLists.txt (hunk 1 above)
Comment (take-cheeze) [reply chainerx_cc/CMakeLists.txt]: OK

File: chainerx_cc/CMakeLists.txt [hunk 3]
```diff
@@ -65,15 +73,67 @@ if(NOT ((CMAKE_BUILD_TYPE STREQUAL "Release") OR (CMAKE_BUILD_TYPE STREQUAL "Deb
...
+
+#--------------
+# Dependencies
+#--------------
 
+# CUDA
+if(${CHAINERX_BUILD_CUDA})
     find_package(CUDA QUIET)
     if(${CUDA_FOUND})
+        list(APPEND CUDA_NVCC_FLAGS -std=c++14)
```
Comment (take-cheeze) [inline chainerx_cc/CMakeLists.txt]: I'd like this more:
```suggestion
        list(APPEND CUDA_NVCC_FLAGS "-std=c++${CMAKE_CXX_STANDARD}")
```

Comment (niboshi) [reply chainerx_cc/CMakeLists.txt]: Thank you for suggestion. Applied.

**Comment by niboshi** [comment]
BTW, the previous change to move `list(APPEND CUDA_NVCC_FLAGS -std=c++14)` was mistakenly amended into the merge commit. I fixed that by separating commits (now it's ac47a0df88af391f18508a9f8dc94498b1a0152a).

File: chainerx_cc/CMakeLists.txt (hunk 2 above)
Comment (take-cheeze) [reply chainerx_cc/CMakeLists.txt]: `CUDA_NVCC_FLAGS` seems optimal for now since current cmake requirement is 3.1

**Comment by take-cheeze** [comment]
Jenkins, test this please!

**Comment by take-cheeze** [comment]
Jenkins, test this please!

**Comment by take-cheeze** [comment]
Test failure seems to be unrelated so merging

**Pull request merged**